src/lib/types/MessageUpdate.ts

```diff
@@ -1,46 +1,110 @@
-import type { WebSearchSource } from "./WebSearch";
+import type { WebSearch, WebSearchSource } from "$lib/types/WebSearch";
+import type { ToolCall, ToolResult } from "$lib/types/Tool";
 
-export type FinalAnswer = {
-	type: "finalAnswer";
-	text: string;
-};
+export type MessageUpdate =
+	| MessageStatusUpdate
+	| MessageTitleUpdate
+	| MessageToolUpdate
+	| MessageWebSearchUpdate
+	| MessageStreamUpdate
+	| MessageFileUpdate
+	| MessageFinalAnswerUpdate;
 
-export type TextStreamUpdate = {
-	type: "stream";
-	token: string;
-};
+export enum MessageUpdateType {
+	Status = "status",
+	Title = "title",
+	Tool = "tool",
+	WebSearch = "webSearch",
+	Stream = "stream",
+	File = "file",
+	FinalAnswer = "finalAnswer",
+}
 
-export type AgentUpdate = {
-	type: "agent";
-	agent: string;
-	content: string;
-	binary?: Blob;
-};
+// Status
+export enum MessageUpdateStatus {
+	Started = "started",
+	Error = "error",
+	Finished = "finished",
+}
+export interface MessageStatusUpdate {
+	type: MessageUpdateType.Status;
+	status: MessageUpdateStatus;
+	message?: string;
+}
 
-export type WebSearchUpdate = {
-	type: "webSearch";
-	messageType: "update" | "error" | "sources";
+// Web search
+export enum MessageWebSearchUpdateType {
+	Update = "update",
+	Error = "error",
+	Sources = "sources",
+	Finished = "finished",
+}
+export interface MessageWebSearchErrorUpdate {
```

Comment:
for `MessageWebSearchErrorUpdate, MessageWebSearchGeneralUpdate, MessageWebSearchFinishedUpdate, ...`, we can probably do:
```ts
interface BaseMessageWebSearch<T=MessageUpdateType> {
    type: T,
    subtype,
    ...
}

type MessageWebSearchErrorUpdate = BaseMessageWebSearch<MessageUpdateType.WebSearch> & { args?: string[] }
...
```